File: android/app/src/main/kotlin/dev/deliteai/assistant/presentation/components/VoiceOverlay.kt
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.deliteai.assistant.presentation.components

import dev.deliteai.assistant.presentation.ui.theme.accent
import dev.deliteai.assistant.presentation.ui.theme.accentHigh2
import dev.deliteai.assistant.presentation.viewmodels.ChatViewModel
import dev.deliteai.assistant.utils.VoiceOverlayState
import androidx.compose.animation.animateColor
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.animation.core.updateTransition
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.ripple
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun VoiceOverlay(chatViewModel: ChatViewModel) {
    LaunchedEffect(Unit) {
        chatViewModel.getLLMAudioFromVoiceInput()
    }
    LaunchedEffect(Unit) {
        return@LaunchedEffect
    }
    val normalizedVolume = ((chatViewModel.volumeState.floatValue + 120) / 120).coerceIn(0f, 1f)
    val currentState = when {
        chatViewModel.isUserSpeaking.value -> VoiceOverlayState.SPEAKING
        else -> VoiceOverlayState.IDLE
    }

    val view = LocalView.current
    DisposableEffect(view) {
        view.keepScreenOn = true
        onDispose { view.keepScreenOn = false }
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = 0.8f))
    ) {
        Column(
            modifier = Modifier
                .align(Alignment.Center),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            AnimatedVoiceOrb(
                voiceState = currentState,
                normalizedVolume = normalizedVolume,
                baseSize = 120.dp,
                chatViewModel
            )
            Spacer(modifier = Modifier.height(16.dp))
            AnimatedSpeechText(
                isUserSpeaking = chatViewModel.isUserSpeaking.value,
                currentText = chatViewModel.spokenText.value,
                persistedText = chatViewModel.persistedRecognizedText.value
            )
            Spacer(modifier = Modifier.height(20.dp))
            if (chatViewModel.isInterruptButtonVisible.value) {
                Text("Interrupt", modifier = Modifier.clickable {
                    chatViewModel.getLLMAudioFromVoiceInput()
                })
            }
        }
        IconButton(
            onClick = {
                chatViewModel.cancelLLMAndClearAudioQueue()
                chatViewModel.isOverlayVisible.value = false
            },
            modifier = Modifier.align(Alignment.TopEnd)
        ) {
            Icon(Icons.Default.Close, contentDescription = "Close")
        }
    }
}

object VoiceOrbAnimations {
    @Composable
    fun getOrbTransition(voiceState: VoiceOverlayState) =
        updateTransition(targetState = voiceState, label = "OrbTransition")

    @Composable
    fun getIdlePulse() = rememberInfiniteTransition().animateFloat(
        initialValue = 0.8f,
        targetValue = 0.95f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1500, easing = FastOutSlowInEasing),
            repeatMode = RepeatMode.Reverse
        )
    )

    @Composable
    fun getInfiniteRotation() = rememberInfiniteTransition().animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 1200, easing = LinearOutSlowInEasing),
            repeatMode = RepeatMode.Restart
        )
    )
}

@Composable
fun AnimatedVoiceOrb(
    voiceState: VoiceOverlayState,
    normalizedVolume: Float,
    baseSize: Dp,
    chatViewModel: ChatViewModel
) {
    val transition = VoiceOrbAnimations.getOrbTransition(voiceState)
    val idlePulse by VoiceOrbAnimations.getIdlePulse()
    val scaleForStates by transition.animateFloat(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "scaleAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> 1f
            VoiceOverlayState.SPEAKING -> 1f + normalizedVolume * 0.3f
        }
    }
    val scale = if (voiceState == VoiceOverlayState.IDLE) idlePulse else scaleForStates
    val orbAlpha by transition.animateFloat(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "orbAlphaAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> 0.8f
            VoiceOverlayState.SPEAKING -> 1f
        }
    }
    val loadingArcAlpha by transition.animateFloat(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "loadingArcAlphaAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> 0f
            VoiceOverlayState.SPEAKING -> 0f
        }
    }
    val startColor by transition.animateColor(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "startColorAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> accent
            VoiceOverlayState.SPEAKING -> Color(0xFF972A2A)
        }
    }
    val midColor by transition.animateColor(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "midColorAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> accentHigh2
            VoiceOverlayState.SPEAKING -> Color(0xFFD35B5B)
        }
    }
    val endColor by transition.animateColor(
        transitionSpec = { tween(durationMillis = 500, easing = FastOutSlowInEasing) },
        label = "endColorAnim"
    ) { state ->
        when (state) {
            VoiceOverlayState.IDLE -> Color.White
            VoiceOverlayState.SPEAKING -> Color(0xFFEBC0C0)
        }
    }
    val rotation by VoiceOrbAnimations.getInfiniteRotation()
    val colorRotation by rememberInfiniteTransition().animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 4000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        )
    )
    val waveAlpha by rememberInfiniteTransition().animateFloat(
        initialValue = 0.3f,
        targetValue = 0.8f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 2000, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse
        )
    )
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(baseSize)
            .scale(scale)
    ) {
        drawOrbShadow(orbAlpha, startColor, endColor)
        drawRotatingLoadingArc(loadingArcAlpha, startColor, rotation)
        drawSphereSurface(orbAlpha, startColor, midColor, endColor, colorRotation, chatViewModel)
        drawOuterWave(waveAlpha, listOf(startColor, midColor, endColor))
    }
}

@Composable
private fun BoxScope.drawOrbShadow(alpha: Float, c1: Color, c2: Color) {
    Canvas(modifier = Modifier.fillMaxSize()) {
        if (alpha > 0f) {
            drawCircle(
                brush = Brush.radialGradient(
                    colors = listOf(
                        c1.copy(alpha = alpha * 0.3f),
                        c2.copy(alpha = 0f)
                    ),
                    center = center,
                    radius = size.minDimension * 1.6f
                ),
                center = center
            )
        }
    }
}

@Composable
private fun BoxScope.drawRotatingLoadingArc(alpha: Float, color: Color, rotation: Float) {
    Canvas(modifier = Modifier.fillMaxSize()) {
        if (alpha > 0f) {
            rotate(rotation) {
                drawArc(
                    color = color.copy(alpha = alpha),
                    startAngle = 0f,
                    sweepAngle = 120f,
                    useCenter = false,
                    style = Stroke(
                        width = size.minDimension / 10f,
                        cap = StrokeCap.Round
                    )
                )
            }
        }
    }
}

@Composable
private fun BoxScope.drawSphereSurface(
    alpha: Float,
    c1: Color,
    c2: Color,
    c3: Color,
    rotation: Float,
    chatViewModel: ChatViewModel
) {
    Canvas(
        modifier = Modifier
            .fillMaxSize()
            .clip(CircleShape)
            .clickable(
                onClick = { chatViewModel.shouldRestartListening() },
                interactionSource = remember { MutableInteractionSource() },
                indication = ripple(color = Color.DarkGray.copy(alpha = 0.3f))
            )
    ) {
        if (alpha > 0f) {
            rotate(rotation) {
                drawCircle(
                    brush = Brush.linearGradient(
                        colors = listOf(
                            c1.copy(alpha = alpha),
                            c2.copy(alpha = alpha),
                            c3.copy(alpha = alpha)
                        ),
                        start = Offset(0f, center.y),
                        end = Offset(size.width, center.y)
                    ),
                    radius = size.minDimension / 2f
                )
            }
        }
    }
}

@Composable
private fun BoxScope.drawOuterWave(alpha: Float, colors: List<Color>) {
    Canvas(modifier = Modifier.fillMaxSize()) {
        drawCircle(
            brush = Brush.radialGradient(
                colors = listOf(
                    colors.first().copy(alpha = alpha * 0.2f),
                    colors.last().copy(alpha = alpha * 0.1f)
                )
            ),
            radius = size.minDimension / 1.3f
        )
    }
}

@Composable
fun AnimatedSpeechText(
    isUserSpeaking: Boolean,
    currentText: String,
    persistedText: String
) {
    val textToShow = when {
        isUserSpeaking -> "Listening..."
        currentText.isNotBlank() -> currentText
        persistedText.isNotBlank() -> persistedText
        else -> "Tap to Speak! Ask me anything..."
    }
    Box(
        modifier = Modifier
            .padding(16.dp)
            .fillMaxWidth(),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = textToShow,
            textAlign = TextAlign.Center,
            color = Color.White,
            style = MaterialTheme.typography.bodyMedium.copy(
                fontSize = 14.sp,
                fontWeight = FontWeight.Medium
            )
        )
    }
}

File: ios/NimbleEdgeAI/views/VoiceOverlay.swift
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import SwiftUI
import AVFAudio
import Combine
enum VoiceOverlayState {
    case idle
    case speaking
}

struct VoiceOverlay: View {
    @ObservedObject var chatViewModel: ChatViewModel
    @StateObject private var speechRecognizer = SpeechRecognizer()
    @State var voiceState: VoiceOverlayState = .speaking
    let onDismiss: () -> Void
    
    var body: some View {
        ZStack {
            Color.black.opacity(0.8)
                .ignoresSafeArea()
                .onTapGesture {
                    onDismiss()
                }
            
            VStack(spacing: 16) {
                AnimatedVoiceOrb(
                    voiceState: $voiceState,
                    normalizedVolume: normalizedVolume,
                    baseSize: 120,
                    chatViewModel: chatViewModel,
                    speechRecognizer: speechRecognizer
                )
                
                VStack {
                    
                    Text(
                        speechRecognizer.isRecording ? "Listening..." :
                        chatViewModel.isLLMSpeaking ? speechRecognizer.transcript :
                        "Tap to Speak! Ask me anything..."
                    )
                    .foregroundColor(.white)
                    .font(.system(size: 14, weight: .medium))
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 16)
                }
                .frame(maxWidth: .infinity)
                .padding(16)
                
                if chatViewModel.isLLMSpeaking {
                    Text("Interrupt")
                        .foregroundColor(.white)
                        .onTapGesture {
                            chatViewModel.cancelTTS()
                        }
                }
            }
            VStack {
                HStack {
                    Spacer()
                    Button(action: {
                        speechRecognizer.stopRecording()
                        chatViewModel.isOverlayVisible = false
                        onDismiss()
                    }) {
                        Image(systemName: "xmark")
                            .foregroundColor(.white)
                            .font(.title2)
                    }
                    .padding()
                }
                Spacer()
            }
        }
        .preferredColorScheme(.dark)
        .onChange(of: speechRecognizer.isRecording, perform: { newValue in
            voiceState = newValue ? .speaking : .idle
        })
        .onChange(of: chatViewModel.isLLMSpeaking, perform: { newValue in
            if chatViewModel.isLLMSpeaking == false {
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1, execute: {
                    speechRecognizer.startRecording()
                })
            }
        })
        .onAppear {
            DispatchQueue.main.async(execute: {
                speechRecognizer.startRecording()
            })
            
            speechRecognizer.onRecordingStoped = {
                let finalTranscript = speechRecognizer.transcript
                chatViewModel.addNewMessageToChatHistory(message: finalTranscript, isUserInput: true)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.2, execute: {
                    print("speechRecognizer.transcript ->> \(finalTranscript)")
                    chatViewModel.passTextInputToLLM(finalTranscript)
                })
            }
        }
        .onDisappear {
            if speechRecognizer.isRecording {
                speechRecognizer.stopRecording()
            }
        }
    }
    
    private var normalizedVolume: Float {
        return ((chatViewModel.volumeState + 120) / 120).clamped(to: 0...1)
    }
    
    private var currentState: VoiceOverlayState {
        return speechRecognizer.isRecording ? .speaking : .idle
    }
}

struct AnimatedVoiceOrb: View {
    @Binding var voiceState: VoiceOverlayState
    let normalizedVolume: Float
    let baseSize: CGFloat
    @ObservedObject var chatViewModel: ChatViewModel
    @ObservedObject var speechRecognizer: SpeechRecognizer
    
    @State private var idlePulse: CGFloat = 0.8
    @State private var rotation: Double = 0
    @State private var colorRotation: Double = 0
    @State private var waveAlpha: Double = 0.3
    
    var body: some View {
        ZStack {
            OuterWaveView(
                alpha: waveAlpha,
                colors: waveColors,
                size: baseSize
            )

            OrbShadowView(
                alpha: orbAlpha,
                startColor: startColor,
                endColor: endColor,
                size: baseSize
            )
            
            if loadingArcAlpha > 0 {
                RotatingLoadingArc(
                    alpha: loadingArcAlpha,
                    color: startColor,
                    rotation: rotation,
                    size: baseSize
                )
            }
            
            SphereSurfaceView(
                alpha: orbAlpha,
                startColor: startColor,
                midColor: midColor,
                endColor: endColor,
                rotation: colorRotation,
                size: baseSize,
                onTap: {
                    if speechRecognizer.isRecording {
                        speechRecognizer.stopRecording()
                        chatViewModel.addNewMessageToChatHistory(message: speechRecognizer.transcript, isUserInput: true)
                        chatViewModel.passTextInputToLLM(speechRecognizer.transcript)
                    } else {
                        speechRecognizer.startRecording()
                    }
                }
            )
        }
        .frame(width: baseSize, height: baseSize)
        .scaleEffect(currentScale)
        .onAppear {
            startAnimations()
        }
    }
    
    private var currentScale: CGFloat {
        switch voiceState {
        case .idle:
            return idlePulse
        case .speaking:
            return 1.0 + CGFloat(normalizedVolume) * 0.3
        }
    }
    
    private var orbAlpha: Double {
        switch voiceState {
        case .idle: return 0.8
        case .speaking: return 1.0
        }
    }
    
    private var loadingArcAlpha: Double {
        switch voiceState {
        case .idle: return 0.0
        case .speaking: return 0.0
        }
    }
    
    private var startColor: Color {
        switch voiceState {
        case .idle: return Color.accent
        case .speaking: return Color(hex: 0xFF972A2A)
        }
    }
    
    private var midColor: Color {
        switch voiceState {
        case .idle: return Color.accentHigh2
        case .speaking: return Color(hex: 0xFFD35B5B)
        }
    }
    
    private var endColor: Color {
        switch voiceState {
        case .idle: return Color.white
        case .speaking: return Color(hex: 0xFFEBC0C0)
        }
    }
    
    private var waveColors: [Color] {
        return [startColor, midColor, endColor]
    }
    
    private func startAnimations() {
        withAnimation(
            Animation.easeInOut(duration: 1.5)
                .repeatForever(autoreverses: true)
        ) {
            idlePulse = 0.95
        }
        
        withAnimation(
            Animation.linear(duration: 1.2)
                .repeatForever(autoreverses: false)
        ) {
            rotation = 360
        }
        
        withAnimation(
            Animation.linear(duration: 4.0)
                .repeatForever(autoreverses: false)
        ) {
            colorRotation = 360
        }
        
        withAnimation(
            Animation.linear(duration: 2.0)
                .repeatForever(autoreverses: true)
        ) {
            waveAlpha = 0.8
        }
    }
}

struct OrbShadowView: View {
    let alpha: Double
    let startColor: Color
    let endColor: Color
    let size: CGFloat
    
    var body: some View {
        if alpha > 0 {
            Circle()
                .fill(
                    RadialGradient(
                        colors: [
                            startColor.opacity(alpha * 0.3),
                            endColor.opacity(0)
                        ],
                        center: .center,
                        startRadius: 0,
                        endRadius: size * 0.8
                    )
                )
                .frame(width: size * 1.6, height: size * 1.6)
        }
    }
}

struct RotatingLoadingArc: View {
    let alpha: Double
    let color: Color
    let rotation: Double
    let size: CGFloat
    
    var body: some View {
        if alpha > 0 {
            Circle()
                .trim(from: 0, to: 0.33)
                .stroke(
                    color.opacity(alpha),
                    style: StrokeStyle(
                        lineWidth: size / 10,
                        lineCap: .round
                    )
                )
                .frame(width: size, height: size)
                .rotationEffect(.degrees(rotation))
        }
    }
}

struct SphereSurfaceView: View {
    let alpha: Double
    let startColor: Color
    let midColor: Color
    let endColor: Color
    let rotation: Double
    let size: CGFloat
    let onTap: () -> Void
    
    var body: some View {
        if alpha > 0 {
            Button(action: onTap) {
                Circle()
                    .fill(
                        LinearGradient(
                            colors: [
                                startColor.opacity(alpha),
                                midColor.opacity(alpha),
                                endColor.opacity(alpha)
                            ],
                            startPoint: .leading,
                            endPoint: .trailing
                        )
                    )
                    .frame(width: size, height: size)
            }
            .rotationEffect(.degrees(rotation))
            .buttonStyle(PlainButtonStyle())
        }
    }
}

struct OuterWaveView: View {
    let alpha: Double
    let colors: [Color]
    let size: CGFloat
    
    var body: some View {
        Circle()
            .fill(
                RadialGradient(
                    colors: [
                        colors.first?.opacity(alpha * 0.2) ?? Color.clear,
                        colors.last?.opacity(alpha * 0.1) ?? Color.clear
                    ],
                    center: .center,
                    startRadius: 0,
                    endRadius: size / 1.3
                )
            )
            .frame(width: size * 1.3, height: size * 1.3)
    }
}
